overlord/snapstate/snapmgr.go

```diff
@@ -405,22 +549,40 @@ func (m *SnapManager) doLinkSnap(t *state.Task, _ *tomb.Tomb) error {
 }
 
 func (m *SnapManager) undoLinkSnap(t *state.Task, _ *tomb.Tomb) error {
-	t.State().Lock()
+	st := t.State()
+
+	// Hold the lock for the full duration of the task here so
+	// nobody observes a world where the state engine and
+	// the file system are reporting different things.
+	st.Lock()
+	defer st.Unlock()
+
 	ss, snapst, err := snapSetupAndState(t)
-	t.State().Unlock()
 	if err != nil {
 		return err
 	}
 
 	// No need to undo "snaps" in state here. The only chance of
 	// having the new state there is a working doLinkSnap call.
-	newDir := ss.MountDir()
-	oldDir := ""
-	if len(snapst.Sequence) > 0 {
-		latest := snapst.Sequence[len(snapst.Sequence)-1]
-		oldDir = snap.MountDir(ss.Name, latest.Revision)
+
+	// relinking of the old snap is done in the undo of unlink-current-snap
+
+	newInfo, err := retrieveInfo(ss.Name, snapst.Candidate)
```

Comment:
This will explode as we've set Candidate to nil in doLinkSnap. To undo it, I think we'll need to first move the tail of the Sequence back into Candidate.